projects/core/src/routing/event/routing-event.builder.ts

```diff
@@ -0,0 +1,164 @@
+      this.buildProductDetailsPageVisitedEvent()
+    );
+    this.eventService.register(
+      CategoryPageVisited,
+      this.buildCategoryPageVisitedEvent()
+    );
+    this.eventService.register(
+      HomePageVisited,
+      this.buildHomePageVisitedEvent()
+    );
+    this.eventService.register(CartPageVisited, this.buildCartVisitedEvent());
+    this.eventService.register(PageVisited, this.buildPageVisitedEvent());
+    this.eventService.register(
+      OrderConfirmationPageVisited,
+      this.orderConfirmationPageVisitedEvent()
+    );
+  }
+
+  protected buildPageVisitedEvent(): Observable<PageVisited> {
+    return this.getNavigatedEvent().pipe(
+      map((state) => createFrom(PageVisited, state))
+    );
+  }
+
+  protected buildHomePageVisitedEvent(): Observable<HomePageVisited> {
+    return this.getCurrentPageContextFor('home').pipe(
+      map((pageContext) => createFrom(HomePageVisited, pageContext))
+    );
+  }
+
+  protected buildProductDetailsPageVisitedEvent(): Observable<
+    ProductDetailsPageVisited
+  > {
+    return this.getCurrentPageContextFor('product').pipe(
+      map((context) => context.id),
+      switchMap((productId) => {
+        return this.productService.get(productId).pipe(
+          filter(Boolean),
+          map((product) => {
+            return createFrom(ProductDetailsPageVisited, product);
+          })
+        );
+      })
+    );
+  }
+
+  protected buildCartVisitedEvent(): Observable<CartPageVisited> {
+    return this.getCurrentPageContextFor('cart').pipe(
+      map((pageContext) => createFrom(CartPageVisited, pageContext))
+    );
+  }
+
+  protected orderConfirmationPageVisitedEvent(): Observable<
+    OrderConfirmationPageVisited
+  > {
+    return this.getCurrentPageContextFor('orderConfirmation').pipe(
+      map((pageContext) =>
+        createFrom(OrderConfirmationPageVisited, pageContext)
+      )
+    );
+  }
+
+  //FIXME: see https://github.com/SAP/spartacus/issues/7991
+  protected buildCategoryPageVisitedEvent(): Observable<CategoryPageVisited> {
+    return this.productSearchService.getResults().pipe(
+      filter((searchResults) => {
+        return (
+          searchResults.breadcrumbs && searchResults.breadcrumbs.length > 0
+        );
+      }),
+      withLatestFrom(this.getCurrentPageContextFor('category')),
+      filter(
+        ([searchResults, _pageContext]) =>
+          searchResults.breadcrumbs && searchResults.breadcrumbs.length > 0
+      ),
+      map(([searchResults, pageContext]) => ({
+        categoryCode: pageContext.id,
+        categoryName: searchResults.breadcrumbs[0].facetValueName,
+      })),
+      map((categoryPage) => createFrom(CategoryPageVisited, categoryPage))
+    );
+  }
+
+  protected searchResultPageVisited(): Observable<SearchPageVisited> {
+    return this.productSearchService.getResults().pipe(
+      filter((searchResults) => Boolean(searchResults.breadcrumbs)),
+      withLatestFrom(this.getCurrentPageContextFor('search')),
```

Comment:
Maybe I'm missing something but... IMO we should emit each time the current page context goes 'search' and then append the latest snapshot of the search results:
```typescript
    return this.getCurrentPageContextFor('search').pipe(,
      withLatestFrom(
         this.productSearchService.getResults().pipe(
             filter((searchResults) => Boolean(searchResults.breadcrumbs))
         )
      ),
```
Why it's implemented the other way around?

Reply:
It appears this has been moved to another PR